### DISCUSSION TITLE
403 errors when trying to save changes from a M2M interface extension

### DISCUSSION THREAD

**Discussion by robertpetermansulzer** [directus/directus#9372] [Archive]
I'm getting 403 errors when trying to save changes made to a collection item from a m2m interface extension. This happens on both rc.99 and 100 although its easier to test on 99 because rc.100 seems to have issues assigning custom interfaces to m2m fields (separate issue).

The out of the box m2m interface from Directus has no problems when changes are saved using that.

I've verified no hooks or permissions set through the admin panel are impacting this. I've also been testing from an admin account.

### To Reproduce

I have a simple project in a docker project which can be viewed here - https://github.com/robertpetermansulzer/directus-m2m-bug-report
This docker instance uses rc.99 instead of 100 for reasons mentioned above.

To reproduce
1. Clone the repo and cd in
2. Run docker commands for setting up DB

- docker-compose -f bug-report-db.yaml pull
- docker-compose -f bug-report-db.yaml up -d
 
 3. Run commands for directus
 
-  docker-compose -f bug-report-directus.yaml pull
-  docker-compose -f bug-report-directus.yaml up -d
 
 4. cd into `m2m-interface-extension` and do a yarn install. Then yarn run dev.
 
 5. Restart directus instance to bring in the extension that was just built.
 
 6. Make two collections. Add a field to collection one with a M2M relation with collection 2. Check off the "add field to other collection" box. Assign the m2m interface extension to the field (comes up as "test name"). Add name fields to each collection so they have some sort of handle.
 
 7.  Add an item to collection 2. If this is the first item of the collection it should have an id of 1. Go to collection one - click the "Test" button - which sets the m2m field value to [1] - and click save. The 403 error should pop up. 

### Errors Shown

POST http://localhost:8055/items/collection_one 403 (Forbidden)
Same for PUT

Same error show ups in the UI.

**Comment by rijkvanzanten** [comment]
Moving this to a discussion, as per your comments:

> I'm getting 403 errors when trying to save changes made to a collection item from a m2m interface extension.

> The out of the box m2m interface from Directus has no problems when changes are saved using that.

There's no special sauce in the API to handle core interfaces differently or anything, so it must be a difference in how you're sending up data to the API.

What's the payload you're sending up to the API when you're staging values from your custom interface?

**Reply by robertpetermansulzer** [reply]
I've tried a few different formats but to keep things simple I'm sending an array of integers. In the example's case its [1] which I intend to reference the related collection item with an id of one.

**Reply by rijkvanzanten** [reply]
Do all the primary keys in your array exist? The 403 could indicate that the item doesn't exist

**Reply by robertpetermansulzer** [reply]
yea I've tried with existing ones and non existing and I get the same result.

**Reply by rijkvanzanten** [reply]
In a m2m, the array of integers are the primary keys of the _junction rows_. When you say "yea I've tried with existing ones and non existing", did you use junction rows or the related rows?

**Reply by robertpetermansulzer** [reply]
Ok that might be where I'm going wrong - the values I've been sending are not primary keys from the junction rows. I'm taking the key of the actual item from the collection that I want to associate to.

**Reply by rijkvanzanten** [reply]
Ah gotcha, that's most likely where the issue is coming from then 👍🏻 

If you look at the payload from the regular m2m interface, you might notice it sends up

```json
[
  {
    "related_id": 1
  },
  {
    "related_id": 2
  },
  {
    "related_id": 3
  }
]
```

instead of 

```json
[1, 2, 3]
```

Those nested objects are the junction rows that will be created 👍🏻  (Make sure to substitute `related_id` with the correct m2o field name)

**Reply by robertpetermansulzer** [reply]
ok great - thanks for the assistance on that, much appreciated! I'll give that a shot.
